clients/src/main/java/org/apache/kafka/clients/producer/internals/ProducerBatch.java

```diff
@@ -152,10 +160,20 @@ private boolean tryAppendForSplit(long timestamp, ByteBuffer key, ByteBuffer val
     public void done(long baseOffset, long logAppendTime, RuntimeException exception) {
         log.trace("Produced messages to topic-partition {} with base offset offset {} and error: {}.",
                   topicPartition, baseOffset, exception);
+        FinalState finalState = exception != null ? FinalState.FAILED : FinalState.SUCCEEDED;
+        if (!this.finalState.compareAndSet(null, finalState)) {
+            if (this.finalState.get() == FinalState.ABORTED) {
+                log.debug("ProduceResponse returned for {} after batch had already been aborted.", topicPartition);
+                return;
```

Comment:
Yeah, not much else we can do since we've already signaled the future and invoked callbacks. I think it would be justifiable to elevate the log level to info if that helps.